_source/_posts/2022-02-25-thymeleaf-security.md

```diff
@@ -0,0 +1,670 @@
+
+
+{% img blog/thymeleaf-security/home-authenticated.png alt:"Home Page Authenticated" width:"800" %}{: .center-image }
+
+
+## Protect content areas with authorization
+
+Let's add a `userProfile.html` template that will display the claims contained in the access token returned from Okta, and also the authorities that spring security derives from the token.
+
+```html
+<html xmlns:th="http://www.thymeleaf.org">
+<head>
+    <title>User Details</title>
+    <!--/*/ <th:block th:include="head :: head"/> /*/-->
+</head>
+<body id="samples">
+  <div th:replace="menu :: menu"></div>
+
+    <div id="content" class="container">
+
+      <div>
+        <h2>My Profile</h2>
+        <p>Hello, <span th:text="${#authentication.principal.attributes['name']}">Joe Coder</span>. Below is the information that was read with your <a href="https://developer.okta.com/docs/api/resources/oidc.html#get-user-information">Access Token</a>.
+        </p>
+        <p>This route is protected with the annotation <code>@PreAuthorize("hasAuthority('SCOPE_profile')")</code>, which will ensure that this page cannot be accessed until you have authenticated, and have the scope <code>profile</code>.</p>
+      </div>
+
+
+      <table class="table table-striped">
+        <thead>
+          <tr>
+            <th>Claim</th>
+            <th>Value</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr th:each="item : ${details}">
+                <td th:text="${item.key}">Key</td>
+                <td th:id="${'claim-' + item.key}" th:text="${item.value}">Value</td>
+            </tr>
+        </tbody>
+      </table>
+
+      <table class="table table-striped">
+        <thead>
+        <tr>
+          <th>Spring Security Authorities</th>
+        </tr>
+        </thead>
+        <tbody>
+        <tr th:each="scope : ${#authentication.authorities}">
+          <td><code th:text="${scope}">Authority</code></td>
+        </tr>
+        </tbody>
+      </table>
+
+    </div>
+  </body>
+<!--/*/ <th:block th:include="footer :: footer"/> /*/-->
+</html>
+```
+
+Add the route mapping in the `HomeController`:
+
+```java
+@GetMapping("/profile")
+@PreAuthorize("hasAuthority('SCOPE_profile')")
+public Mono<Rendering> userDetails(OAuth2AuthenticationToken authentication) {
+    return Mono.just(Rendering.view("userProfile").modelAttribute("details", authentication.getPrincipal().getAttributes())
+            .build());
+}
+```
+
+The annotation `@PreAuthorize` allows to define an authorization predicate that can be written in SpEL (Spring Expression Language), and will be checked before the method execution. Only users with the authority `SCOPE_PROFILE` can request a display of the `userProfile` page. This is server-side protection.
```

Comment:
```suggestion
The `@PreAuthorize` annotation allows to define an authorization predicate that can be written in SpEL (Spring Expression Language). It will be checked before the method execution. Only users with the authority `SCOPE_PROFILE` can request a display of the `userProfile` page. This is server-side protection.
```